crawl-ref/source/mon-act.cc

```diff
@@ -2292,18 +2279,28 @@ static void _post_monster_move(monster* mons)
     }
 
     if (mons->has_ench(ENCH_RING_OF_THUNDER))
-    {
-        // TODO: deduplicate with mon-ench.cc
-        cloud_type ctype = CLOUD_STORM;
+        surround_mons_with_cloud(mons, CLOUD_STORM);
 
-        for (adjacent_iterator ai(mons->pos()); ai; ++ai)
-            if (!cell_is_solid(*ai)
-                && (!cloud_at(*ai)
-                    || cloud_at(*ai)->type == ctype))
-            {
-                place_cloud(ctype, *ai, 2 + random2(3), mons);
-            }
-    }
+    if (mons->has_ench(ENCH_RING_OF_FLAMES))
+        surround_mons_with_cloud(mons, CLOUD_FIRE);
+
+    if (mons->has_ench(ENCH_RING_OF_CHAOS))
+        surround_mons_with_cloud(mons, CLOUD_CHAOS);
+
+    if (mons->has_ench(ENCH_RING_OF_MUTATION))
+        surround_mons_with_cloud(mons, CLOUD_MUTAGENIC);
+
+    if (mons->has_ench(ENCH_RING_OF_FOG))
+        surround_mons_with_cloud(mons, CLOUD_GREY_SMOKE);
+
+    if (mons->has_ench(ENCH_RING_OF_ICE))
+        surround_mons_with_cloud(mons, CLOUD_COLD);
+
+    if (mons->has_ench(ENCH_RING_OF_DRAINING))
+        surround_mons_with_cloud(mons, CLOUD_NEGATIVE_ENERGY);
+
+    if (mons->has_ench(ENCH_RING_OF_ACID))
+        surround_mons_with_cloud(mons, CLOUD_ACID);
```

Comment:
This should be replaced with a `static map<mon_ench_type, cloud_type>` that you iterate over to pick the cloud type and update. That iteration logic should be in its own function in `mon-ench.cc` called something like `update_mons_cloud_ring` and used below as well.